src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1965 @@
+        .is_leader_slot = false,
+        .prev_leader_slot = null,
+        .slot_vote_tracker = null,
+        .cluster_slot_pubkeys = null,
+        .total_epoch_stake = 0,
+    };
+
+    pub fn deinit(self: PropagatedStats, allocator: std.mem.Allocator) void {
+        var propagated_validators = self.propagated_validators;
+        propagated_validators.deinit(allocator);
+
+        var propagated_node_ids = self.propagated_node_ids;
+        propagated_node_ids.deinit(allocator);
+
+        if (self.slot_vote_tracker) |svt| svt.arc_ed.rwlock_ed.deinit(allocator);
+        var maybe_csp = if (self.cluster_slot_pubkeys) |csp| csp.arc_ed.rwlock_ed else null;
+        if (maybe_csp) |*csp| csp.deinit(allocator);
+    }
+
+    pub fn clone(
+        self: PropagatedStats,
+        allocator: std.mem.Allocator,
+    ) std.mem.Allocator.Error!PropagatedStats {
+        var propagated_validators = try self.propagated_validators.clone(allocator);
+        errdefer propagated_validators.deinit(allocator);
+
+        var propagated_node_ids = try self.propagated_node_ids.clone(allocator);
+        errdefer propagated_node_ids.deinit(allocator);
+
+        const maybe_slot_vote_tracker: ?cluster_info_vote_listener.SlotVoteTracker = blk: {
+            const svt = self.slot_vote_tracker orelse break :blk null;
+            break :blk try svt.arc_ed.rwlock_ed.clone(allocator);
+        };
+        errdefer if (maybe_slot_vote_tracker) |svt| svt.deinit(allocator);
+
+        var maybe_cluster_slot_pubkeys: ?cluser_slots_service.SlotPubkeys = blk: {
+            const csp = self.cluster_slot_pubkeys orelse break :blk null;
+            break :blk try csp.arc_ed.rwlock_ed.clone(allocator);
+        };
+        errdefer if (maybe_cluster_slot_pubkeys) |*svt| svt.deinit(allocator);
+
+        return .{
+            .propagated_validators = propagated_validators,
+            .propagated_node_ids = propagated_node_ids,
+            .propagated_validators_stake = self.propagated_validators_stake,
+            .is_propagated = self.is_propagated,
+            .is_leader_slot = self.is_leader_slot,
+            .prev_leader_slot = self.prev_leader_slot,
+            .slot_vote_tracker = if (maybe_slot_vote_tracker) |svt| .{
+                .arc_ed = .{ .rwlock_ed = svt },
+            } else null,
+            .cluster_slot_pubkeys = if (maybe_cluster_slot_pubkeys) |csp| .{
+                .arc_ed = .{ .rwlock_ed = csp },
+            } else null,
+            .total_epoch_stake = self.total_epoch_stake,
+        };
+    }
+
+    /// Same as `addVotePubkeyAssumeCapacity`, except it [re-]allocates if needed, instead of
+    /// assuming capacity is sufficient.
+    pub fn addVotePubkey(
+        self: *PropagatedStats,
+        allocator: std.mem.Allocator,
+        vote_pubkey: Pubkey,
+        stake: u64,
+    ) std.mem.Allocator.Error!bool {
+        const gop = try self.propagated_validators.getOrPut(allocator, vote_pubkey);
+        gop.value_ptr.* = {};
```

Comment:
Does this actually do anything?

Reply:
It doesn't do anything, it just makes it clear to a reader of the code that the value is simply a void value; omitting it would make it seem like at a glance that there's a missing initialization for `gop.value_ptr.*`, until you look deeper and see that it's a `Map(Pubkey, void)`.